### PR TITLE
check prior fork consistently

### PR DESCRIPTION
**Pull request by zhiqiangxu**
I see [this check](https://github.com/ethereum-optimism/optimism/blob/d48b45954c381f75a13e61312da68d84e9b41418/op-chain-ops/genesis/config.go#L394-L402) on `DeployConfig`, if it holds, then the same should hold for fork times inside the rollup config since initially these fork times are [generated](https://github.com/ethereum-optimism/optimism/blob/d48b45954c381f75a13e61312da68d84e9b41418/op-chain-ops/genesis/config.go#L593-L598) from `DeployConfig`.

This check keeps users from manually modifying the fork times inconsistently after initially generated from `DeployConfig`.

### CODE REVIEW

File: op-node/node/config.go [hunk 1]
```diff
@@ -138,6 +138,18 @@ func (cfg *Config) Check() error {
 		if err := cfg.Beacon.Check(); err != nil {
 			return fmt.Errorf("misconfigured L1 Beacon API endpoint: %w", err)
 		}
+		if cfg.Rollup.CanyonTime == nil {
```
Comment (tynes) [inline op-node/node/config.go]: This all should go inside of `cfg.Rollup.Check()` so that everwhere that calls `Check` gets this

Comment (zhiqiangxu) [reply op-node/node/config.go]: Good point, done!

File: op-node/rollup/types.go [hunk 2]
```diff
@@ -277,6 +277,20 @@ func (cfg *Config) Check() error {
 	if cfg.L2ChainID.Sign() < 1 {
 		return ErrL2ChainIDNotPositive
 	}
+	if cfg.EcotoneTime != nil {
```
Comment (trianglesphere) [inline op-node/rollup/types.go]: Could we instead use a function like `checkForkComesAfter(first, second *uint64)` and then run call it for the following pairs: `(regolith, canyon), (canyon, delta), (delta, ecotone)`. We're going to be adding more forks in the future & it needs to scale well.

Also we tend to return named errors & it'd be nice to add some unit tests for this check.

**Review by trianglesphere — COMMENTED** [review]
Thanks for picking up this ticket. There's a little bit more work to get this merged.

File: op-node/rollup/types.go (hunk 2 above)
Comment (zhiqiangxu) [reply op-node/rollup/types.go]: Thanks for the advice, I basically copied the `checkFork` function from `DeployConfig`, not sure whether it's good enough or we'd like to extract it to some common place? Waiting for your advice for the next step:)

Comment (trianglesphere) [reply op-node/rollup/types.go]: let's pull it to a top level (but private function). The usage looks great

Comment (zhiqiangxu) [reply op-node/rollup/types.go]: Done:)

**Review by trianglesphere — COMMENTED** [review]
Code LGTM, last thing is to add a test case for these additions. We have scaffolding here: https://github.com/ethereum-optimism/optimism/blob/58fab6b324aaf32513338e8c9b68fbec49e2ba67/op-node/rollup/types_test.go#L291

**Comment by zhiqiangxu** [comment]
> Code LGTM, last thing is to add a test case for these additions. We have scaffolding here:
> 
> https://github.com/ethereum-optimism/optimism/blob/58fab6b324aaf32513338e8c9b68fbec49e2ba67/op-node/rollup/types_test.go#L291

I added some test cases for forks in another loop since the error returned by `checkFork` is not a named one.

**Review by trianglesphere — APPROVED** [review]
LGTM. Thanks for fixing this.

**Review by 0x00101010 — APPROVED** [review]

**Review by tynes — APPROVED** [review]

**Comment by tynes** [comment]
Looks like this is stuck on `shell-check` - could this be due to renaming the job?

**Comment by ajsutton** [comment]
It very likely doesn't have the rename in this branch but the job is now required - pull in the latest changes from develop and it should be good to go.

**Comment by zhiqiangxu** [comment]
I'm seeing such errors from `op-e2e-WS-tests`: `L2 safe head ahead of L2 unsafe head` / `Nil finalized block cannot evict old blobs`. Not familiar with `op-e2e` yet, is this a random environment error ?

**Pull request merged**